src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderCursor.cs

```diff
@@ -464,9 +467,69 @@ public LineBatch GetBatch()
                     throw Contracts.ExceptDecode(batch.Exception, "Stream reading encountered exception");
                 }
 
+                private static class MultiLineReader
+                {
+                    // When reading lines that contain quoted fields, the quoted fields can contain
+                    // '\n' so we we'll need to read multiple lines (multilines) to get all the fields
+                    // of a given row.
+                    public static string ReadMultiLine(TextReader sr, StringBuilder sb, bool ignoreHashLine)
+                    {
+                        string line;
+                        line = sr.ReadLine();
+
+                        // if it was an empty line or if we've reached the end of file (i.e. line = null)
+                        if (string.IsNullOrEmpty(line))
+                            return line;
+
+                        // In ML.NET we filter out lines beginning with // and # at the beginning of the file
+                        // Or lines beginning with // elsewhere in the file.
+                        // Thus, we don't care to check if there's a quoted multiline when the line begins with
+                        // these chars.
+                        if (line[0] == '/' && line[1] == '/')
+                            return line;
+                        if (ignoreHashLine && line[0] == '#')
+                            return line;
+
+                        // Get more lines until the number of quote characters is even
+                        // 2 consecutive quotes are considered scaped quotes
+                        long numOfQuotes = GetNumberOfChars(line, '"');
+                        if (numOfQuotes % 2 == 0)
+                            return line;
+
+                        sb.Clear();
+                        sb.Append(line);
+                        while (numOfQuotes % 2 != 0)
+                        {
+                            line = sr.ReadLine();
+
+                            if (line == null) // If we've reached the end of the file
+                                break; // MYTODO: This could happen if we have an invalid open quote which never closes so we reach the end of the file without properly closing the field, should we throw instead in this case?
```

Comment:
Should we throw in this case?

Reply:
I am not sure throwing is a good idea. If they are parsing a large file and they see an exception, they won't be able to proceed with the experiment at all. 
What about ignoring the whole line and adding a warning (e.g. End of file encountered while searching for matching quote"

I don't see infrastructure in here to log a warning. We would need to add that and also work with ModelBuilder to surface the warnings appropriately?

---
In reply to: [425034529](https://github.com/dotnet/machinelearning/pull/5125#discussion_r425034529) [](ancestors = 425034529)

Reply:
> What about ignoring the whole line and adding a warning (e.g. End of file encountered while searching for matching quote"
I can't do that when reading multilines precisely because that is the behavior we're accepting; e.g. if I have the following row:
```
11,"this is a
multiline
field",dog
```
And if we have reading multilines enabled, then there's no way to know if the new line character after `"this is a` is supposed to be a format error where we don't find the matching quote, or if it's actually a multiline. In that case, it is a correct multiline, as accepted by the RFC, but  the following example wouldn't be correctly formatted:
```
11,"this is a
multiline
field,dog
```
and it would actually read the rest of the file until another doublequote (") is found (which could potentially be all the rest of the file).

> I don't see infrastructure in here to log a warning. We would need to add that and also work with ModelBuilder to surface the warnings appropriately?
Even without multilines (i.e. even without this PR) we actually already support reading and logging poorly formated lines. E.g., the following line is unacceptable if allowQuoting is enabled, but readMultilines isn't:
```
11,"this is a mistake,dog
12,"this is not",cat
```
When encountering this, the TextLoader will actually log this in [here](https://github.com/dotnet/machinelearning/blob/a2406f64c96bb2e5bd7e7c9ba91487e31d673d6b/src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderParser.cs#L942-L943) and will return the line as having empty fields (see other occurrences of ParseStats.LogBadFmt and ParseStats.LogBadValue to see other scenarios where logging occurs). Returning empty fields is desired because we don't want to stop loading a dataset only because a parsing error.

On the other hand, if there's a problem transforming a field into a value (e.g. having `"this shouldn't be a string", "this should", horse`, when the first column is supposed to be an int), then an exception is thrown (for example in [here](https://github.com/dotnet/machinelearning/blob/a2406f64c96bb2e5bd7e7c9ba91487e31d673d6b/src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderParser.cs#L1349)) informing the line that caused the problem. ModelBuilder actually does show an error message box including this exception if it occurs.

So in general I don't think there's work to do on this direction.

---
In reply to: [425461605](https://github.com/dotnet/machinelearning/pull/5125#discussion_r425461605) [](ancestors = 425461605,425034529)

Reply:
The case you point out above is when ReadLine returns null. In that case you have run out of things to read and it is the last line in the file.  If you have a mismatched quote in this case, is there anything you can do apart from logging the issue and warning the user or throwing an exception? 

---
In reply to: [425471411](https://github.com/dotnet/machinelearning/pull/5125#discussion_r425471411) [](ancestors = 425471411,425461605,425034529)

Reply:
I don't think there's anything else we can do except logging or throwing an exception. We could simply ignore it in here, but maybe in the TextLoader.Parser it will actually log or throw an exception if it can't parse that line (which is very likely). In any case, we can't avoid loading more than necessary (and potentially load the whole dataset into memory) if the CSV file contains this mistake.

---
In reply to: [425500545](https://github.com/dotnet/machinelearning/pull/5125#discussion_r425500545) [](ancestors = 425500545,425471411,425461605,425034529)

Reply:
Might want to look into how other ML and CSV-reader packages handle the run-away case. E.g. pandas, numpy, databricks, excel, ...

Reply:
So I've decided to actually throw an exception in here. In general, it shouldn't be that common for a user to actually hit this exception, so I think it's safe to put it.

After playing around with Excel, it doesn't support very well having an open quote that doesn't close, and it tends to mix up columns.

In pandas, sometimes it throws errors, such as in:
```
>>> data = 'num,text\n5,"hello, \\"Bob\\", nice to see you\n6,Hi Alice'
>>> pd.read_csv(StringIO(data), escapechar='\\')

[...]
pandas.errors.ParserError: Error tokenizing data. C error: EOF inside string starting at row 1
```

And other times it doesn't parse it correctly, messing up the type of column (the first column is integer, but got inferred as `object` (i.e. `string`)... without the syntax error in the CSV file it actually gets correctly inferred as int):
```
>>> data = 'num,text\n5,"hello, \\"Bob\\", nice to see you\n6,"Hi,Alice"'
>>> df = pd.read_csv(StringIO(data), escapechar='\\')
>>> df
                                   num                                 text
5  hello, "Bob", nice to see you\n6,Hi        Alice"
>>> df.dtypes
num     object
text    object
dtype: object
```

Since in ML.NET we don't infer column types, both of the above cases will result in an exception during parsing even if the exception isn't thrown here in the LineReader. Nonetheless, there might be some weird cases where no exception is thrown on the parser when this problem occurs, so I think it's better to have an early exception to make sure this case fails. Notice that, in any case, it's somewhat uncommon to actually hit this place, so I think that an exception is good enough without having to worry to much about it.

---
In reply to: [425753156](https://github.com/dotnet/machinelearning/pull/5125#discussion_r425753156) [](ancestors = 425753156)

Reply:
_Future_: We may also want a parameter for max row length (default ~500MB). That can limit the OOM possibility of run-away lines. I'm worried that we'll read in 1TB and throw at the very end of the file.

Reply:
Ok. I'll open an issue once this PR gets merged to track this.

---
In reply to: [427002585](https://github.com/dotnet/machinelearning/pull/5125#discussion_r427002585) [](ancestors = 427002585)